jsonrpc/debug_handler.go

```diff
@@ -0,0 +1,179 @@
+package jsonrpc
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"regexp"
+	"runtime/pprof"
+	"runtime/trace"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/go-bexpr"
+)
+
+var (
+	errCPUProfilingInProgress    = errors.New("CPU profiling already in progress")
+	errCPUProfilingNotInProgress = errors.New("CPU profiling not in progress")
+	errTraceAlreadyInProgress    = errors.New("trace already in progress")
+	errTraceNotInProgress        = errors.New("trace not in progress")
+	identifierRegex              = regexp.MustCompile(`[:/\.A-Za-z0-9_-]+`)
+	notRegex                     = regexp.MustCompile(`!([:/\.A-Za-z0-9_-]+)`)
+)
+
+// DebugHandler implements the debugging API.
+// Do not create values of this type, use the one
+// in the Handler variable instead.
+type DebugHandler struct {
+	mux            sync.Mutex
+	cpuW           io.WriteCloser
+	cpuProfileFile string
+	traceW         io.WriteCloser
+	traceFile      string
+}
+
+// StartCPUProfile turns on CPU profiling, writing to the given file.
+func (debug *DebugHandler) StartCPUProfile(file string) error {
+	debug.mux.Lock()
+	defer debug.mux.Unlock()
+
+	if debug.cpuW != nil {
+		return errCPUProfilingInProgress
+	}
+
+	f, err := os.Create(expandHome(file))
+	if err != nil {
+		return err
+	}
+
+	if err := pprof.StartCPUProfile(f); err != nil {
+		f.Close()
+
+		return err
+	}
+
+	debug.cpuW = f
+	debug.cpuProfileFile = file
+
+	return nil
+}
+
+// StopCPUProfile stops an ongoing CPU profile.
+func (debug *DebugHandler) StopCPUProfile() error {
+	debug.mux.Lock()
+	defer debug.mux.Unlock()
+
+	if debug.cpuW == nil {
+		return errCPUProfilingNotInProgress
+	}
+
+	pprof.StopCPUProfile()
+
+	debug.cpuW.Close()
+	debug.cpuW = nil
```

Comment:
Propagate the error?

```suggestion
	if err := debug.cpuW.Close(); err != nil {
		return err
	}
```